File: excelium-core/src/main/java/excelium/core/writer/AbstractWorkbookWriter.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.core.writer;

import org.apache.commons.collections4.CollectionUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Abstract workbook writer.
 * This class implements some helper methods for writing cell values.
 *
 * @author PhungDucKien
 * @since 2018.04.29
 */
public abstract class AbstractWorkbookWriter implements WorkbookWriter {

    /**
     * Flush values
     */
    protected Map<String, List<List<Object>>> flushValues = new HashMap<>();

    @Override
    public void setCellValue(Object value, String cell) throws IOException {
        Map<String, Object> valueMap = new HashMap<>();
        valueMap.put(cell, value);
        batchSetCellValues(valueMap, cell);
    }

    @Override
    public void setRowCellValues(List<Object> values, String row) throws IOException {
        Map<String, List<Object>> valueMap = new HashMap<>();
        valueMap.put(row, values);
        batchSetRowCellValues(valueMap, row);
    }

    @Override
    public void setColumnCellValues(List<Object> values, String column) throws IOException {
        Map<String, List<Object>> valueMap = new HashMap<>();
        valueMap.put(column, values);
        batchSetColumnCellValues(valueMap, column);
    }

    @Override
    public void setRangeCellValues(List<List<Object>> values, String range) throws IOException {
        Map<String, List<List<Object>>> valueMap = new HashMap<>();
        valueMap.put(range, values);
        batchSetRangeCellValues(valueMap, range);
    }

    @Override
    public void batchSetCellValues(Map<String, Object> values, String... cells) throws IOException {
        Map<String, List<List<Object>>> valueMap = new HashMap<>();
        for (String cell : cells) {
            List<List<Object>> rangeValues = new ArrayList<>();
            List<Object> rowValues = new ArrayList<>();
            rowValues.add(values.get(cell));
            rangeValues.add(rowValues);
            valueMap.put(cell, rangeValues);
        }
        batchSetRangeCellValues(valueMap, cells);
    }

    @Override
    public void batchSetRowCellValues(Map<String, List<Object>> values, String... rows) throws IOException {
        Map<String, List<List<Object>>> valueMap = new HashMap<>();
        for (String row : rows) {
            List<List<Object>> rangeValues = new ArrayList<>();
            rangeValues.add(values.get(row));
            valueMap.put(row, rangeValues);
        }
        batchSetRangeCellValues(valueMap, rows);
    }

    @Override
    public void batchSetColumnCellValues(Map<String, List<Object>> values, String... columns) throws IOException {
        Map<String, List<List<Object>>> valueMap = new HashMap<>();
        for (String column : columns) {
            List<List<Object>> rangeValues = new ArrayList<>();
            List<Object> columnValues = values.get(column);
            if (CollectionUtils.isNotEmpty(columnValues)) {
                for (Object value : columnValues) {
                    List<Object> rowValues = new ArrayList<>();
                    rowValues.add(value);
                    rangeValues.add(rowValues);
                }
            }
            valueMap.put(column, rangeValues);
        }
        batchSetRangeCellValues(valueMap, columns);
    }

    @Override
    public void batchSetRangeCellValues(Map<String, List<List<Object>>> values, String... ranges) throws IOException {
        for (String range : ranges) {
            if (values.containsKey(range)) {
                flushValues.put(range, values.get(range));
            }
        }
    }

    @Override
    public void close() throws IOException {
        if (flushValues != null && !flushValues.isEmpty()) {
            flush();
        }
    }
}

File: packages/core/src/writer/AbstractWorkbookWriter.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CollectionUtil } from '@excelium/common';
import WorkbookWriter from './WorkbookWriter';

/**
 * Abstract workbook writer.
 * This class implements some helper methods for writing cell values.
 *
 * @author PhungDucKien
 * @since 2018.04.29
 */
export default abstract class AbstractWorkbookWriter implements WorkbookWriter {
  /**
   * Flush values
   */
  protected flushValues = new Map<string, any[][]>();

  public setCellValue(value: any, cell: string): void {
    const valueMap = new Map<string, any>();
    valueMap.set(cell, value);
    this.batchSetCellValues(valueMap, [cell]);
  }

  public setRowCellValues(values: any[], row: string): void {
    const valueMap = new Map<string, any[]>();
    valueMap.set(row, values);
    this.batchSetRowCellValues(valueMap, [row]);
  }

  public setColumnCellValues(values: any[], column: string): void {
    const valueMap = new Map<string, any[]>();
    valueMap.set(column, values);
    this.batchSetColumnCellValues(valueMap, [column]);
  }

  public setRangeCellValues(values: any[][], range: string): void {
    const valueMap = new Map<string, any[][]>();
    valueMap.set(range, values);
    this.batchSetRangeCellValues(valueMap, [range]);
  }

  public batchSetCellValues(values: Map<string, any>, cells: string[]): void {
    const valueMap = new Map<string, any[][]>();
    for (const cell of cells) {
      const rangeValues = [] as any[][];
      const rowValues = [] as any[];
      rowValues.push(values.get(cell));
      rangeValues.push(rowValues);
      valueMap.set(cell, rangeValues);
    }
    this.batchSetRangeCellValues(valueMap, cells);
  }

  public batchSetRowCellValues(values: Map<string, any[]>, rows: string[]): void {
    const valueMap = new Map<string, any[][]>();
    for (const row of rows) {
      const rangeValues = [] as any[][];
      const value = values.get(row);
      if (value && CollectionUtil.isNotEmpty(value)) {
        rangeValues.push(value);
      }
      valueMap.set(row, rangeValues);
    }
    this.batchSetRangeCellValues(valueMap, rows);
  }

  public batchSetColumnCellValues(values: Map<string, any[]>, columns: string[]): void {
    const valueMap = new Map<string, any[][]>();
    for (const column of columns) {
      const rangeValues = [] as any[][];
      const columnValues = values.get(column);
      if (columnValues && CollectionUtil.isNotEmpty(columnValues)) {
        for (const value of columnValues) {
          const rowValues = [] as any[];
          rowValues.push(value);
          rangeValues.push(rowValues);
        }
      }
      valueMap.set(column, rangeValues);
    }
    this.batchSetRangeCellValues(valueMap, columns);
  }

  public batchSetRangeCellValues(values: Map<string, any[][]>, ranges: string[]): void {
    for (const range of ranges) {
      const value = values.get(range);
      if (value && CollectionUtil.isNotEmpty(value)) {
        this.flushValues.set(range, value);
      }
    }
  }

  public close(): Promise<void> {
    if (this.flushValues && this.flushValues.size > 0) {
      return this.flush();
    }
    return Promise.resolve();
  }

  public abstract flush(): Promise<void>;
}
